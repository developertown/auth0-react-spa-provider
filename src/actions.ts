import Auth0Client from "@auth0/auth0-spa-js/dist/typings/Auth0Client";
import actionCreatorFactory from "typescript-fsa";

import { TokenUser } from "./types";

const createAction = actionCreatorFactory("AUTH0_PROVIDER");

export type Auth0LoadedParams = {
  auth0Client: Auth0Client;
  user?: TokenUser;
  handleRedirectCallback: Auth0Client["handleRedirectCallback"];
  loginWithPopup: Auth0Client["loginWithPopup"];
};
export const auth0Loaded = createAction<Auth0LoadedParams>("LOADED");

export type LoginWithPopupActionResult = {
  user?: TokenUser;
};
export const loginWithPopupAction = createAction.async<void, LoginWithPopupActionResult>("LOGIN_WITH_POPUP");

export type HandleRedirectCallbackResult = {
  user?: TokenUser;
};
export const handleRedirectCallbackAction = createAction.async<void, HandleRedirectCallbackResult>(
  "HANDLE_REDIRECT_CALLBACK",
);

export type Auth0ProviderAction =
  | ReturnType<typeof auth0Loaded>
  | ReturnType<typeof loginWithPopupAction.started>
  | ReturnType<typeof loginWithPopupAction.done>
  | ReturnType<typeof loginWithPopupAction.failed>
  | ReturnType<typeof handleRedirectCallbackAction.started>
  | ReturnType<typeof handleRedirectCallbackAction.done>
  | ReturnType<typeof handleRedirectCallbackAction.failed>;
