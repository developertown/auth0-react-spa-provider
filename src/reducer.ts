import { reducerWithoutInitialState } from "typescript-fsa-reducers";

import { auth0Loaded, handleRedirectCallbackAction, loginWithPopupAction } from "./actions";
import {
  Auth0ProviderState,
  AuthenticatedState,
  HandlingRedirectState,
  isHandlingRedirectState,
  isLoadedState,
  isLoadingState,
  LoadingState,
  UnauthenticatedState,
} from "./types";

export const auth0ProviderStateReducer = reducerWithoutInitialState<Auth0ProviderState>()
  .case(
    auth0Loaded.started,
    (state): LoadingState => {
      return {
        ...state,
        loading: true,
      } as LoadingState;
    },
  )
  .case(auth0Loaded.done, (state, { result: { user, auth0Client, handleRedirectCallback, loginWithPopup } }):
    | AuthenticatedState
    | UnauthenticatedState => {
    if (!isLoadingState(state)) {
      throw new Error("Invalid state: auth0Loaded may only be applied when in the loading state.");
    }

    return {
      ...state,

      auth0Client,

      loginWithRedirect: auth0Client.loginWithRedirect.bind(auth0Client),
      getTokenWithPopup: auth0Client.getTokenWithPopup.bind(auth0Client),
      getTokenSilently: auth0Client.getTokenSilently.bind(auth0Client),
      getIdTokenClaims: auth0Client.getIdTokenClaims.bind(auth0Client),
      logout: auth0Client.logout.bind(auth0Client),

      handleRedirectCallback,
      loginWithPopup,

      loading: false,
      popupOpen: false,
      user: user || undefined,
      isAuthenticated: Boolean(user),
    } as AuthenticatedState | UnauthenticatedState;
  })
  .case(loginWithPopupAction.started, (state) => {
    if (!isLoadedState(state)) {
      throw new Error("Invalid state: loginWithPopupAction.started may only be applied when in a loaded state.");
    }

    return {
      ...state,
      popupOpen: true,
    };
  })
  .case(loginWithPopupAction.done, (state, { result: user }): AuthenticatedState | UnauthenticatedState => {
    if (!isLoadedState(state)) {
      throw new Error("Invalid state: loginWithPopupAction.done may only be applied when in a loaded state.");
    }

    return {
      ...state,
      popupOpen: false,
      user: user || undefined,
      isAuthenticated: Boolean(user),
    } as AuthenticatedState | UnauthenticatedState;
  })
  .case(handleRedirectCallbackAction.started, (state) => {
    if (!isLoadedState(state)) {
      throw new Error(
        "Invalid state: handleRedirectCallbackAction.started may only be applied when in a loaded state.",
      );
    }

    return {
      ...state,
      handlingRedirect: true,
    } as HandlingRedirectState;
  })
  .case(handleRedirectCallbackAction.done, (state, { result: user }): AuthenticatedState | UnauthenticatedState => {
    if (!isHandlingRedirectState(state)) {
      throw new Error(
        "Invalid state: handleRedirectCallbackAction.done may only be applied when in the handlingRedirect state.",
      );
    }

    return {
      ...state,
      user,
      isAuthenticated: Boolean(user),
      handlingRedirect: false,
    } as AuthenticatedState | UnauthenticatedState;
  })
  .build();
