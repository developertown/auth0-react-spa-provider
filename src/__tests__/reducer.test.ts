import * as Auth0 from "@auth0/auth0-spa-js";
import { auth0Loaded, handleRedirectCallbackAction, loginWithPopupAction } from "../actions";
import { auth0ProviderStateReducer } from "../reducer";
import {
  HandlingRedirectState,
  isAuthenticatedState,
  isLoadedState,
  isUnauthenticatedState,
  LoadingState,
  TokenUser,
  UnauthenticatedState,
} from "../types";

jest.mock("@auth0/auth0-spa-js");

const createFakeAuth0Client = (): Auth0.Auth0Client =>
  new Auth0.Auth0Client({
    domain: "",
    // eslint-disable-next-line @typescript-eslint/camelcase
    client_id: "",
  });

const createInitialState = (): LoadingState => ({
  loading: true,
  isAuthenticated: false,
  popupOpen: false,
  user: undefined,
  auth0Client: undefined,
  handlingRedirect: false,
});

const createFakeUnauthenticatedState = (
  auth0Client: Auth0.Auth0Client = createFakeAuth0Client(),
): UnauthenticatedState => ({
  isAuthenticated: false,
  loading: false,
  loginWithPopup: jest.fn(),
  handleRedirectCallback: jest.fn(),
  handlingRedirect: false,
  user: undefined,
  popupOpen: false,
  auth0Client,
  loginWithRedirect: auth0Client.loginWithRedirect.bind(auth0Client),
  getTokenWithPopup: auth0Client.getTokenWithPopup.bind(auth0Client),
  getTokenSilently: auth0Client.getTokenSilently.bind(auth0Client),
  getIdTokenClaims: auth0Client.getIdTokenClaims.bind(auth0Client),
  logout: auth0Client.logout.bind(auth0Client),
});

const createHandlingRedirectState = (): HandlingRedirectState =>
  ({
    ...createFakeUnauthenticatedState(),
    handlingRedirect: true,
  } as HandlingRedirectState);

describe("reducer", () => {
  describe("loginWithPopupAction", () => {
    describe("started", () => {
      it("flips the popupOpen flag to true", () => {
        const origState = createFakeUnauthenticatedState();
        const newState = auth0ProviderStateReducer(origState, loginWithPopupAction.started());
        expect(newState.popupOpen).toBeTruthy();
      });

      it("throws an error if the auth0 state is not yet fully loaded", () => {
        const origState = createInitialState();
        expect(() =>
          auth0ProviderStateReducer(origState, loginWithPopupAction.started()),
        ).toThrowErrorMatchingInlineSnapshot(
          `"Invalid state: loginWithPopupAction.started may only be applied when in a loaded state."`,
        );
      });
    });

    describe("done", () => {
      it("flips the popupOpen flag to false", () => {
        const origState = {
          ...createFakeUnauthenticatedState(),
          popupOpen: true,
        };
        const newState = auth0ProviderStateReducer(
          origState,
          loginWithPopupAction.done({ result: { user: undefined } }),
        );
        expect(newState.popupOpen).toBeFalsy();
      });

      it("gets back an authenticated state if a user is provided", () => {
        const origState = {
          ...createFakeUnauthenticatedState(),
          popupOpen: true,
        };
        const newState = auth0ProviderStateReducer(origState, loginWithPopupAction.done({ result: { user: {} } }));
        expect(isAuthenticatedState(newState)).toBeTruthy();
      });

      it("throws an error if the auth0 state is not yet fully loaded", () => {
        const origState = createInitialState();
        expect(() =>
          auth0ProviderStateReducer(origState, loginWithPopupAction.done({ result: { user: undefined } })),
        ).toThrowErrorMatchingInlineSnapshot(
          `"Invalid state: loginWithPopupAction.done may only be applied when in a loaded state."`,
        );
      });
    });
  });

  describe("handleRedirectCallbackAction", () => {
    describe("started", () => {
      it("flips the handlingRedirect flag to true", () => {
        const origState = createFakeUnauthenticatedState();
        const newState = auth0ProviderStateReducer(origState, handleRedirectCallbackAction.started());

        expect(isLoadedState(newState) && newState.handlingRedirect).toBeTruthy();
      });

      it("throws an error if the auth0 state is not yet fully loaded", () => {
        const origState = createInitialState();
        expect(() =>
          auth0ProviderStateReducer(origState, handleRedirectCallbackAction.started()),
        ).toThrowErrorMatchingInlineSnapshot(
          `"Invalid state: handleRedirectCallbackAction.started may only be applied when in a loaded state."`,
        );
      });
    });

    describe("done", () => {
      it("flips the handlingRedirect flag to false", () => {
        const origState = createHandlingRedirectState();
        const newState = auth0ProviderStateReducer(
          origState,
          handleRedirectCallbackAction.done({ result: { user: undefined } }),
        );
        expect(newState.handlingRedirect).toBeFalsy();
      });

      it("gets back an authenticated state if a user is provided", () => {
        const origState = createHandlingRedirectState();
        const newState = auth0ProviderStateReducer(
          origState,
          handleRedirectCallbackAction.done({ result: { user: {} } }),
        );
        expect(isAuthenticatedState(newState)).toBeTruthy();
      });

      it("throws an error if the auth0 state is not in the handling redirect state", () => {
        const origState = createFakeUnauthenticatedState();
        expect(() =>
          auth0ProviderStateReducer(origState, handleRedirectCallbackAction.done({ result: { user: undefined } })),
        ).toThrowErrorMatchingInlineSnapshot(
          `"Invalid state: handleRedirectCallbackAction.done may only be applied when in the handlingRedirect state."`,
        );
      });
    });
  });

  describe("auth0Loaded", () => {
    it("sets up the core values properly", () => {
      const origState = createInitialState();

      const user: TokenUser = {};
      const auth0Client = createFakeAuth0Client();
      const handleRedirectCallback = jest.fn();
      const loginWithPopup = jest.fn();

      const newState = auth0ProviderStateReducer(
        origState,
        auth0Loaded.done({
          result: {
            user,
            // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
            // @ts-ignore
            auth0Client,
            handleRedirectCallback,
            loginWithPopup,
          },
        }),
      );

      expect(isLoadedState(newState)).toBeTruthy();
      expect(newState.handlingRedirect).toBeFalsy();
      expect(newState.popupOpen).toBeFalsy();
      expect(newState.handlingRedirect).toBeFalsy();
      expect(newState.loading).toBeFalsy();

      expect(isLoadedState(newState) && newState.auth0Client).toEqual(auth0Client);
      expect(isLoadedState(newState) && newState.handleRedirectCallback).toEqual(handleRedirectCallback);
      expect(isLoadedState(newState) && newState.loginWithPopup).toEqual(loginWithPopup);
    });

    it("returns an authenticated state when appropriate", () => {
      const origState = createInitialState();

      const user: TokenUser = {};
      const auth0Client = createFakeAuth0Client();
      const handleRedirectCallback = jest.fn();
      const loginWithPopup = jest.fn();

      const newState = auth0ProviderStateReducer(
        origState,
        auth0Loaded.done({
          result: {
            user,
            // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
            // @ts-ignore
            auth0Client,
            handleRedirectCallback,
            loginWithPopup,
          },
        }),
      );

      expect(isAuthenticatedState(newState)).toBeTruthy();
    });

    it("returns an unauthenticated state when appropriate", () => {
      const origState = createInitialState();

      const user = undefined;
      const auth0Client = createFakeAuth0Client();
      const handleRedirectCallback = jest.fn();
      const loginWithPopup = jest.fn();

      const newState = auth0ProviderStateReducer(
        origState,
        auth0Loaded.done({
          result: {
            user,
            // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
            // @ts-ignore
            auth0Client,
            handleRedirectCallback,
            loginWithPopup,
          },
        }),
      );

      expect(isUnauthenticatedState(newState)).toBeTruthy();
    });
  });
});
