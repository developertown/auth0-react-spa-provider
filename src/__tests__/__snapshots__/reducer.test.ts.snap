// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer handleRedirectCallbackAction done throws an error if the auth0 state is not in the handling redirect state 1`] = `"Invalid state: handleRedirectCallbackAction.done may only be applied when in the handlingRedirect state."`;

exports[`reducer handleRedirectCallbackAction started throws an error if the auth0 state is not yet fully loaded 1`] = `"Invalid state: handleRedirectCallbackAction.started may only be applied when in a loaded state."`;

exports[`reducer loginWithPopupAction done throws an error if the auth0 state is not yet fully loaded 1`] = `"Invalid state: loginWithPopupAction.done may only be applied when in a loaded state."`;

exports[`reducer loginWithPopupAction started throws an error if the auth0 state is not yet fully loaded 1`] = `"Invalid state: loginWithPopupAction.started may only be applied when in a loaded state."`;
